### ISSUE TITLE
cli.Command: Argument order is changed depending on the flag set

### ISSUE THREAD

**Issue by brenol** [codegangsta/cli#355]
Hello!

I'm not sure if this is the expected behavior, however the following code:

```go
package main

import (
	"fmt"
	"os"

	"github.com/codegangsta/cli"
)

func main() {
	app := cli.NewApp()

	app.Commands = []cli.Command{
		cli.Command{
			Name:  "update",
			Usage: "update a report",
			Flags: []cli.Flag{
				cli.StringFlag{
					Name:  "filename",
					Usage: "filename",
				},
			},
			Action: func(c *cli.Context) {
				fmt.Println(c.Args())
			},
		},
	}

	app.Run(os.Args)
}
```

Runs differently depending on the position of the arguments passed:

```bash
$ ./main update Arg1 Arg2 Arg3 --filename test.json
[Arg1 Arg2 Arg3]

$ ./main update Arg1 --filename test.json Arg2 Arg3
[Arg2 Arg3 Arg1]
```

Expected behavior for me would be that when running the second command, the following should output:

```bash
$ ./main update Arg1 --filename test.json Arg2 Arg3
[Arg1 Arg2 Arg3]
```

It seems that the fact that the flag is in the middle is breaking the Args() order.

Thanks

**Comment by jszwedko** [comment]
@brenol definitely a bug, thank you for the report!

**Comment by jszwedko** [comment]
This appears to be due to the argument reordering that is done to try to trick the stdlib `flag` parser to handling commands that have arguments specified before flags (the stdlib `flag` library stops flag parsing as soon as it encounters the first argument). At first blush, I think supporting this behavior would require much more complicated preprocessing of flags, but I'm open to reviewing a PR addressing this. I'm almost more inclined to remove the re-ordering feature as it appears to only work in very specific circumstances (if you specify all of your arguments before you specify flags).

https://github.com/codegangsta/cli/blob/master/command.go#L104-L116

**Comment by jszwedko** [comment]
I think this supersedes #103 too -- my opinion is to remove the reordering.

**Comment by danslimmon** [comment]
Removing the ordering seems like a fine answer. I looked at this for a while today, and there's really no way to fix it without, like you said @jszwedko, a whole bunch of complicated flag preprocessing.

**Comment by jszwedko** [comment]
Fixed by #398 (requires use of the `v2` branch however as the change was not backwards compatible). Let me know what you think!
